### PR TITLE
Customizable tooltip

### PR DESCRIPTION
**Pull request by vaimikpatel2908**
*Added a tooltip parameter to the Tooltip widget to accept custom widget as a parameter to create a custom tooltip like for the info button in the image. It will continue to accept the message parameter so it will not break the current code.*

image

*I could not perform the test as forked repository was having the conflicting dependencies problem. I tried to fix by changing dependency versions to supported versions but it introduced many functions or newly added code not found errors. if someone can help me to fix these issues as I am contributing for the first time*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by vaimikpatel2908** [comment]
@googlebot I signed it

File: packages/flutter/lib/src/material/tooltip.dart [hunk 1]
```diff
@@ -112,11 +113,17 @@ class Tooltip extends StatefulWidget {
...
     this.child,
-  }) : assert(message != null),
-       super(key: key);
+  })  : assert(
+          tooltip != null || message != null,
+          'A non-null message or tooltip must be provided to a Tooltip Widget',
+        ),
+        super(key: key);
+
+  /// The widget to display in the tooltip.
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/tooltip.dart:122]: This doc should also talk about "message" and explain the difference.

File: packages/flutter/lib/src/material/tooltip.dart [hunk 2]
```diff
@@ -112,11 +113,17 @@ class Tooltip extends StatefulWidget {
...
+  final Widget? tooltip;
 
   /// The text to display in the tooltip.
-  final String message;
+  final String? message;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/tooltip.dart:126]: same for this doc

File: packages/flutter/lib/src/material/tooltip.dart [hunk 3]
```diff
@@ -210,7 +217,8 @@ class Tooltip extends StatefulWidget {
   static bool dismissAllToolTips() {
     if (_openedToolTips.isNotEmpty) {
       // Avoid concurrent modification.
-      final List<_TooltipState> openedToolTips = List<_TooltipState>.from(_openedToolTips);
+      final List<_TooltipState> openedToolTips =
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/tooltip.dart:220]: Please revert all these unrelated code formatting changes. The flutter repo does not use dartfmt, see also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting

Comment (goderbauer) [reply packages/flutter/lib/src/material/tooltip.dart:220]: (Here and elsewhere)

File: packages/flutter/lib/src/material/tooltip.dart [hunk 4]
```diff
@@ -420,7 +448,9 @@ class _TooltipState extends State<Tooltip> with SingleTickerProviderStateMixin {
     );
     _entry = OverlayEntry(builder: (BuildContext context) => overlay);
     overlayState.insert(_entry!);
-    SemanticsService.tooltip(widget.message);
+    if (widget.message != null) {
+      SemanticsService.tooltip(widget.message!);
+    }
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/tooltip.dart:453]: The tooltip still needs to be somehow announced to the semantics layer for accessibility purposes even when a widget is given instead of a message. How do you plan to do that?

File: packages/flutter/lib/src/material/tooltip.dart [hunk 5]
```diff
@@ -112,11 +113,17 @@ class Tooltip extends StatefulWidget {
     this.waitDuration,
     this.showDuration,
     this.child,
-  }) : assert(message != null),
-       super(key: key);
+  })  : assert(
+          tooltip != null || message != null,
+          'A non-null message or tooltip must be provided to a Tooltip Widget',
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/tooltip.dart:118]: nit: end with a `.`

File: packages/flutter/lib/src/material/tooltip.dart [hunk 6]
```diff
@@ -597,9 +642,12 @@ class _TooltipOverlay extends StatelessWidget {
     required this.preferBelow,
     this.onEnter,
     this.onExit,
-  }) : super(key: key);
+  })  : assert(tooltip != null || message != null,
+            'Either tooltip or message must not be null'),
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/tooltip.dart:646]: nit: end with a `.`

**Review by goderbauer — COMMENTED** [review]
This PR is missing tests.

Also, I am not convinced that adding the widget option is a great idea since it is not clear to me how to make this accessible to accessibility services. Have you thought about that?

**Comment by goderbauer** [comment]
Looks like the analyzer is also unhappy, see https://github.com/flutter/flutter/pull/85924/checks?check_run_id=2993432717 and there appear to be some test failures? Please take a look at those.

File: packages/flutter/lib/src/material/tooltip.dart (hunk 1 above)
Comment (vaimikpatel2908) [reply packages/flutter/lib/src/material/tooltip.dart:122]: message can be replaced with the semantic label property like Text Widget or any other widgets. In such a way, we can make it optional for developers so that they can provide it if they wish to use it for accessibility.

File: packages/flutter/lib/src/material/tooltip.dart (hunk 2 above)
Comment (vaimikpatel2908) [reply packages/flutter/lib/src/material/tooltip.dart:126]: I just left the message property as it is to run the existing code as it is but can be deprecated and can introduce new semantic label property as discussed above.

File: packages/flutter/lib/src/material/tooltip.dart (hunk 4 above)
Comment (vaimikpatel2908) [reply packages/flutter/lib/src/material/tooltip.dart:453]: We can deprecate the message property and introduce the semanticLabel property like other widgets and use it for accessibility purposes. Also, I was thinking that widgets also have semantics property but not sure how they will react in real-time.

File: packages/flutter/lib/src/material/tooltip.dart (hunk 6 above)
Comment (vaimikpatel2908) [reply packages/flutter/lib/src/material/tooltip.dart]: added dot in new commit

File: packages/flutter/lib/src/material/tooltip.dart (hunk 5 above)
Comment (vaimikpatel2908) [reply packages/flutter/lib/src/material/tooltip.dart]: added dot in new commit

**Comment by goderbauer** [comment]
Can you please revert all the formatting changes in this PR to make it easier to review?

Also, the analyzer appears to be unhappy.

**Comment by vaimikpatel2908** [comment]
Please check out the last two commits for the code review, which do not include any code formatting changes.

File: packages/flutter/lib/src/material/tooltip.dart (hunk 4 above)
Comment (goderbauer) [reply packages/flutter/lib/src/material/tooltip.dart:433]: Deprecating the message property is probably unnecessarily breaking and makes the Tooltip for the simple cases harder to use.

We still need to figure out what to do about the semantics, though.

File: packages/flutter/lib/src/material/tooltip.dart [hunk 7]
```diff
@@ -597,9 +608,12 @@ class _TooltipOverlay extends StatelessWidget {
     required this.preferBelow,
     this.onEnter,
     this.onExit,
-  }) : super(key: key);
+  })  : assert(tooltip != null || message != null,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/tooltip.dart:611]: nit: format this like the other assert on the Tooltip class.

**Comment by goderbauer** [comment]
Also: This PR is missing tests.

**Comment by vaimikpatel2908** [comment]
@goderbauer  
There are some questions coming to my mind, for that I need your suggestions,

1) I understand your concern for not introducing a breaking change, as it is already in use. The message property is used for tooltip as well as semantics. So, why not separate them according to their purposes? 

2) Though it would be breaking change, But ultimately message property is converted to the Text widget then why not keep only message property or tooltip property to accept only widgets as input. 

Just let me know what do you think about these.

I will also add some tests for this PR in the next commit.

**Comment by christopherfujino** [comment]
This PR looks like it was branched off of a different branch from `master`, resulting in a large diff and meaning it will not be mergeable. Please see [tree hygiene](https://github.com/flutter/flutter/wiki/Tree-hygiene) for more details about how to open pull requests to Flutter repositories.
